src/Compilers/Test/Utilities/CSharp/FunctionPointerUtilities.cs

```diff
@@ -128,6 +131,134 @@ static void verifyParameter(ParameterSymbol symbol, MethodSymbol containing)
+                {
+                    Assert.Empty(semanticInfo.CandidateSymbols);
+                    AssertEx.Equal(expectedSymbol, semanticInfo.Symbol.ToTestDisplayString(includeNonNullable: false));
+                    Assert.Equal(CandidateReason.None, semanticInfo.CandidateReason);
+                }
+                else if (expectedSymbolCandidates is object)
+                {
+                    Assert.Null(semanticInfo.Symbol);
+                    Assert.Equal(expectedReason, semanticInfo.CandidateReason);
+                    AssertEx.Equal(expectedSymbolCandidates, semanticInfo.CandidateSymbols.Select(s => s.ToTestDisplayString(includeNonNullable: false)));
+                }
+                else
+                {
+                    Assert.Null(semanticInfo.Symbol);
+                    Assert.Empty(semanticInfo.CandidateSymbols);
+                    Assert.Equal(CandidateReason.None, semanticInfo.CandidateReason);
+                }
+            }
+
+            static void verifyNestedFunctionPointerSyntaxSemanticInfo(SemanticModel model, IFunctionPointerTypeSymbol ptrType, SeparatedSyntaxList<ParameterSyntax> paramSyntaxes)
+            {
+                // https://github.com/dotnet/roslyn/issues/43321 Nullability in type syntaxes that don't have an origin bound node
+                // can differ.
+                var signature = ptrType.Signature;
+                for (int i = 0; i < paramSyntaxes.Count - 1; i++)
+                {
+                    var paramSyntax = paramSyntaxes[i].Type!;
+                    var paramSemanticInfo = model.GetSemanticInfoSummary(paramSyntax);
+                    ITypeSymbol signatureParamType = signature.Parameters[i].Type;
+
+                    Assert.Equal<ISymbol>(signatureParamType, paramSemanticInfo.Type, SymbolEqualityComparer.Default);
+                    Assert.Equal(paramSemanticInfo.Type, paramSemanticInfo.ConvertedType, SymbolEqualityComparer.IncludeNullability);
+
+                    Assert.Equal(CandidateReason.None, paramSemanticInfo.CandidateReason);
+                    Assert.Equal(signatureParamType, paramSemanticInfo.Type, SymbolEqualityComparer.Default);
+                    Assert.Empty(paramSemanticInfo.CandidateSymbols);
+
+                    if (paramSyntax is FunctionPointerTypeSyntax { Parameters: var paramParamSyntaxes })
+                    {
+                        var paramPtrType = (IFunctionPointerTypeSymbol)paramSemanticInfo.Type!;
+                        CommonVerifyFunctionPointer(paramPtrType.GetSymbol());
+                        verifyNestedFunctionPointerSyntaxSemanticInfo(model, paramPtrType, paramParamSyntaxes);
+                    }
+                }
+
+                var returnParam = paramSyntaxes[^1].Type;
+                var returnSemanticInfo = model.GetSemanticInfoSummary(returnParam!);
+
+                Assert.Equal(signature.ReturnType, returnSemanticInfo.Type, SymbolEqualityComparer.Default);
```

Comment:
>Assert.Equal(signature.ReturnType, returnSemanticInfo.Type, SymbolEqualityComparer.Default); [](start = 16, length = 92)

It looks like there is fair amount of code duplication (with the body of the loop above) from here to the end of the function.